### DISCUSSION TITLE
Is it possible to use mermaid-js inside a quarto file?

### DISCUSSION THREAD

**Discussion by aborruso** [quarto-dev/quarto-cli#669] [Q&A] [answered]
Hi,
it could be very useful to have [mermaid markdown flow-chart](https://mermaid-js.github.io/mermaid/#%2Fflowchart%3Fid=graph) inside quarto.

Is it possible to do it?

Thank you

**Comment by jjallaire** [comment]
Not yet, but we are working on this feature right now!

**Reply by aborruso** [reply]
Great thing, thank you

**Comment by radovan-miletic** [comment]
Hi, here goes my workaround (with R-4.1.3).
I have installed the development version of {DiagrammeR} from GitHub using the {devtools} package:
````
devtools::install_github("rich-iannone/DiagrammeR")
````
Then, [deployed](https://mermaid-js.github.io/mermaid/#/)

> mermaid without a bundler, one can insert a `script` tag with an absolute address and a `mermaidAPI` call into the HTML like so:

and placed `script` tags before diagrammer chunk:
````
<script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
<script>mermaid.initialize({startOnLoad:true});</script>

```{r}
#| label: fig-diagram
#| fig-cap: Diagram
#| echo: false  
DiagrammeR::mermaid(diagram = 'scripts/diagram.mmd', height = '100%', width = '100%')
```
````

> Doing so will command the mermaid parser to look for the `<div>` tags with `class="mermaid"`. From these tags mermaid will try to read the diagram/chart definitons and render them into svg charts.

Hope this would be of some help.

**Reply by aborruso** [reply]
Thank you. I wait it to be made available simply by writing something like

~~~
```mermaid
flowchart TD
    A-->B & C
```
~~~

to have

```mermaid
flowchart TD
    A-->B & C
```

**Reply by radovan-miletic** [reply]
That would be very cool.  

In meanwhile, the script below works for me:
````
<script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
<script>mermaid.initialize({startOnLoad:true});</script>

```{r}
DiagrammeR::mermaid(diagram = '
                  
  flowchart TD
    A-->B & C
  
  ', height = '100%', width = '100%')
```
````

**Reply by jjallaire** [reply]
@aborruso Yes that's more or less the syntax we will use (although we'll use the braced variation of `{mermaid}` for consistency w/ R/Python/OJS). That also means you'll be able to do this:

````markdown
```{mermaid}
%%| echo: false
%%| fig-cap: "Caption"

flowchart TD
    A-->B & C
```
````

**Reply by aborruso** [reply]
Wow, I will wait for it.

Thank you again

**Reply by cscheid** [reply]
Preliminary support for this is now on `main`. This file:

    ---
    title: foo
    format: html
    ---
    
    # This is a test
    
    ```{mermaid}
    %%| echo: false
    %%| fig-cap: "Caption"
    
    flowchart TD
        A-->B & C
    ```
    
    Did it work?

produces this:

<img width="259" alt="image" src="https://user-images.githubusercontent.com/285675/163846567-1cfd7cd6-2ed5-4dce-8f30-aa74c0108c15.png">

There are definitely limitations with the current support. Right now, it only supports producing `html` formats, and it doesn't produce the correct output in `self-contained` mode. If those limitations are not a concern for you, then please go ahead, try it out, and let me know if something isn't supported appropriately. If you need output in `docx` or `pdf` format, then we're looking at about three to four weeks before landing that.

**Reply by aborruso** [reply]
Yeaaa!

Does it works also in reveal.js presentation?

![image](https://user-images.githubusercontent.com/30607/163850313-dbcd6e76-b3a1-40df-b444-f22b70b0b6b5.png)

**Reply by cscheid** [reply, marked as answer]
Yes, it should work.

For the record, the current constraint is that we depend on the mermaid runtime and, as a result, on the existence of a javascript runtime.
